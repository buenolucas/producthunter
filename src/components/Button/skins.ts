import styled, { css, DefaultTheme } from 'styled-components'
import { ButtonProps } from './Button'
import tinycolor from 'tinycolor2'
type ComponentProps = Pick<
  ButtonProps,
  'size' | 'color' | 'fullWidth' | 'outlined'
>

const componentSizeModifiers = {
  small: (theme: DefaultTheme) => css`
    border-radius: ${theme.shape.borderRadius.small};
    font-size: ${theme.typography.button.smSize};
    padding: ${theme.spacing.xxxsmall} ${theme.spacing.xxsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.typography.button.size};
    border-radius: ${theme.shape.borderRadius.medium};
    padding: ${theme.spacing.xxsmall} ${theme.spacing.xsmall};
  `,
  large: (theme: DefaultTheme) => css`
    border-radius: ${theme.shape.borderRadius.large};
    font-size: ${theme.typography.button.lgSize};
    padding: ${theme.spacing.xxsmall} ${theme.spacing.large};
  `,
  fullWidth: () => css`
    width: 100%;
  `
}

export const Component = styled.button<ComponentProps>`
  ${({
    theme: {
      typography: { button }
    },
    theme,
    size,
    fullWidth,
    color,
    outlined
  }) => css`
    font-family: ${button.family};
    font-weight: ${button.weight};
    font-size: ${button.size};
    line-height: ${button.smLineHeight};
    ${!!size && componentSizeModifiers[size](theme)};
    ${!!fullWidth && componentSizeModifiers.fullWidth()};
    ${outlined
      ? css`
          color: ${theme.palette[color!].main};
          border: 1px solid ${theme.palette[color!].main};
          &:hover {
            border-color: ${tinycolor(theme.palette[color!].main)
              .darken(10)
              .toString()};
          }
          &:active {
            border-color: ${tinycolor(theme.palette[color!].main)
              .darken(25)
              .toString()};
          }
        `
      : css`
          background-color: ${theme.palette[color!].main};
          border: none;
          color: ${theme.palette[color!].on};
          &:hover {
            background-color: ${tinycolor(theme.palette[color!].main)
              .darken(10)
              .toString()};
          }
          &:active {
            background-color: ${tinycolor(theme.palette[color!].main)
              .darken(25)
              .toString()};
          }
        `};
  `}
`
