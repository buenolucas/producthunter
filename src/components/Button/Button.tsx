import * as skins from './skins'

export type ButtonProps = {
  children?: React.ReactNode
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  color?: 'blood' | 'groundAndBlood' | 'bootSole'
  outlined?: boolean
  onClick?: () => (event: React.MouseEvent<HTMLButtonElement>) => void
}

const Button = ({
  children,
  size = 'medium',
  color = 'groundAndBlood',
  fullWidth = false,
  outlined = false,
  ...props
}: ButtonProps) => (
  <skins.Component
    data-testid="button-skin"
    size={size}
    color={color}
    fullWidth={fullWidth}
    outlined={outlined}
    {...props}
  >
    <span>{children}</span>
  </skins.Component>
)

export default Button
