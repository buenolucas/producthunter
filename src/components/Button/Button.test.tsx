import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utilities/test/helpers'

import Button from '.'

describe('<Button />', () => {
  it('should render the size and bg for default', () => {
    const { container } = renderWithTheme(<Button>upshot</Button>)

    expect(screen.getByRole('button', { name: /upshot/i })).toHaveStyle({
      'border-radius': '0.8rem',
      'font-size': '2.4rem',
      padding: '0.8rem 1.6rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the small size', () => {
    renderWithTheme(<Button size="small">upshot</Button>)

    expect(screen.getByRole('button', { name: /upshot/i })).toHaveStyle({
      'border-radius': '0.4rem',
      'font-size': '2.2rem',
      padding: '0.4rem 0.8rem'
    })
  })

  it('should render the large size', () => {
    renderWithTheme(<Button size="large">upshot</Button>)

    expect(screen.getByRole('button', { name: /upshot/i })).toHaveStyle({
      'border-radius': '1.0rem',
      'font-size': '3.4rem',
      padding: '0.8rem 4.0rem'
    })
  })

  it('should render a full width version', () => {
    renderWithTheme(<Button fullWidth>upshot</Button>)

    expect(screen.getByRole('button', { name: /upshot/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render blood color button', () => {
    renderWithTheme(<Button color="blood">upshot</Button>)

    expect(screen.getByRole('button', { name: /upshot/i })).toHaveStyle({
      'background-color': '#EB5757'
    })
  })

  it('should render outlined blood color button', () => {
    renderWithTheme(
      <Button color="blood" outlined>
        upshot
      </Button>
    )

    expect(screen.getByRole('button', { name: /upshot/i })).toHaveStyle({
      color: '#EB5757',
      border: '1px solid #EB5757'
    })
  })
})
