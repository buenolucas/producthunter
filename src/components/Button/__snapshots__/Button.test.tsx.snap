// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the size and bg for default 1`] = `
.c0 {
  font-family: 'Smokum',cursive;
  font-weight: 400;
  font-size: 2.4rem;
  font-size: 2.4rem;
  border-radius: 0.8rem;
  padding: 0.8rem 1.6rem;
  background-color: #784321;
  border: none;
  color: #FAFAFA;
}

.c0:hover {
  background-color: #502d16;
}

.c0:active {
  background-color: #140b05;
}

<button
  class="c0"
  color="groundAndBlood"
  data-testid="button-skin"
>
  <span>
    upshot
  </span>
</button>
`
